### ISSUE TITLE
url function is unstoppable (and it is not good:)

### ISSUE THREAD

**Issue by nickitat** [ClickHouse/ClickHouse#66656]
in test `02044_url_glob_parallel_connection_refused` we have the following query:

``` sh
timeout 5s ${CLICKHOUSE_CLIENT} --max_threads 10 --query "SELECT * FROM url('http://128.0.0.{1..10}:${CLICKHOUSE_PORT_HTTP}/?query=SELECT+sleep(1)', TSV, 'x UInt8')" --format Null 2>/dev/null
```

after 5s timeout it will be cancelled, but essentially execution will continue for quite some time:

```
2024-07-17 18:29:39.989302	(from 127.0.0.1:49244) (comment: 02044_url_glob_parallel_connection_refused.sh) SELECT * FROM url(\'http://128.0.0.{1..10}:8123/?query=SELECT+sleep(1)\', TSV, \'x UInt8\') (stage: Complete)
...
2024-07-17 18:29:44.995630	Client has dropped the connection, cancel the query.
```

but then we still do 10 retries and finish only in 3 minutes:

```
...
2024-07-17 18:32:30.799382	Failed to make request to \'http://128.0.0.9:8123/?query=SELECT+sleep(1)\'. Error: \'Timeout: connect timed out: 128.0.0.9:8123\'. Failed at try 10/10.
2024-07-17 18:32:30.800625	Client has gone away.
2024-07-17 18:32:30.800641	Processed in 170.810526814 sec.
```

For now i will create a workaround for the test - lower the number of retries to fit in hung check limit (90s).

**Comment by vitlibar** [comment]
The issue will be fixed in https://github.com/ClickHouse/ClickHouse/pull/104089

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**url function is unstoppable (and it is not good:)**

**ClickGap verified:** reproduced it with the script below on `master`; tested 6 version(s) — 6 affected, 0 clean (see Affects); long-standing (not a recent regression); introducing PR not yet pinned.

### Reproduction
```
-- Point url() at a non-routable host so every connect attempt times out,
-- and ask the query to stop at ~2s via max_execution_time.
SELECT * FROM url('http://192.0.2.1:81/', TSV, 'x UInt8')
SETTINGS http_connection_timeout=1, http_max_tries=6,
         http_retry_initial_backoff_ms=1000, http_retry_max_backoff_ms=2000,
         max_execution_time=2, timeout_overflow_mode='throw';
-- Observed: the query runs the full ~15s retry/back-off sequence instead of
-- stopping at ~2s. With default http_max_tries=10 it runs ~3 minutes. The HTTP
-- retry loop ignores query cancellation, so the read is effectively unstoppable.
```

### Bisect
- Reproduces on every release line tested down to `v24.1.8.22` (the oldest with a downloadable binary for this repro). The introducing change predates `v24.1.8.22`, so it can't be pinned from release binaries — this is a **long-standing** issue, not a recent regression. A maintainer with full git history can `git blame` the relevant file/line to find the exact change.

### Affects
- Verified against `master` build `e692fa12c25c` (aarch64)
- **Reproduces on**: `master, 26.5, 26.4, 26.3, 26.2, 26.1`

### Suggested next step
Manual triage of the `comp-table-functions` path is needed. The reproducer above should be a fast first check.

**Component:** `comp-table-functions`
**Severity:** `P2`

---
_ClickGap · Finding: `h_issue_66656`_




### Triage metadata
- Primary component: `comp-table-functions`
- Secondary components: `comp-query-execution`
- Component owners (`comp-table-functions`): @alexey-milovidov @Avogar
- Possibly fixed by (not yet linked): PR #102801, PR #104089

**Comment by zlareb1** [comment]
Fresh reproduction on master `26.8.1.918` (2026-08-07), connection-refused flavor: with nothing listening on the target port, `SELECT * FROM url('http://localhost:18124/', LineAsString) SETTINGS max_execution_time = 5` spent **130.8 seconds** in the connection-retry backoff before failing with `Poco::Exception ... Connection refused` — 26x the 5-second limit, and the final error is the connection error rather than `TIMEOUT_EXCEEDED`, so the deadline was never consulted during the retry loop. Also observed in a live run as `CancellationChecker: Cancelling the task because of the timeout: 30000 ms` firing while `ReadWriteBufferFromHTTP` kept retrying for ~100 more seconds.

Found by an automatic optimizer-testing framework (differential testing of optimizer settings, query plans, and equivalent rewrites).
